### PR TITLE
Add compile-time flag to remove code for QUIC

### PR DESCRIPTION
**Pull request by syeopite**

### CODE REVIEW

**Comment by syeopite** [comment]
Using Invidious without QUIC has some  issues

**Comment by unixfox** [comment]
Why this got closed? I know how to fix the issues.

**Comment by syeopite** [comment]
There we go, that should fix the gzip issues. However, in order to make Invidious fully functional without lsquic, we're going to need to replace the few dozen usages of `:authority` with a HTTP/1.1 equivalent; otherwise, we can't support images. It's a bit out of scope for this PR though.

File: src/invidious/routes/login.cr
Comment (SamantazFox) [inline src/invidious/routes/login.cr:56]:
```suggestion
          client = CONFIG.use_quic ? QUIC::Client.new(LOGIN_URL) : HTTP::Client.new(LOGIN_URL)
```

File: src/invidious/routes/login.cr
Comment (SamantazFox) [inline src/invidious/routes/login.cr:54]: ditto not required

File: src/invidious/helpers/utils.cr
Comment (SamantazFox) [inline src/invidious/helpers/utils.cr:4]: typo, should be `Connection`

**Review by SamantazFox — CHANGES_REQUESTED** [review]

File: src/invidious/helpers/utils.cr
Comment (syeopite) [reply src/invidious/helpers/utils.cr:67]: It is actually required. The `disable_quic` logic is considered a macro(?) so we can't actually access `conn` unless we declare it outside first. You can try to build Invidious without this yourself, but it'll just result in a compile time error.

Comment (SamantazFox) [reply src/invidious/helpers/utils.cr:67]: Whaaaa, even the inline version doesn't work!
`conn = {% unless flag?(:disable_quic) %} use_quic ? QUIC::Client.new(url) : {% end %} HTTP::Client.new(url)`

This is definitely an upstream bug...

Comment (SamantazFox) [reply src/invidious/helpers/utils.cr:67]: That works, though:

```crystal
conn = (
  {% unless flag?(:disable_quic) %}
    CONFIG.use_quic ? QUIC::Client.new(url) : HTTP::Client.new(url)
  {% else %}
    conn = HTTP::Client.new(url)
  {% end %}
)
```

**Comment by TheFrenchGhosty** [comment]
Blocked by https://github.com/iv-org/invidious/pull/2364#issuecomment-929664683

**Comment by syeopite** [comment]
Running this PR causes the following to be outputted within the logs occasionally. Though, it doesn't seem to actually cause any noticeable effect.


<details>

```
Exception: SSL_connect: I/O error (OpenSSL::SSL::Error)
  from /usr/share/crystal/src/openssl/ssl/socket.cr:34:11 in 'initialize'
  from /usr/share/crystal/src/openssl/ssl/socket.cr:3:5 in 'new:context:sync_close:hostname'
  from src/invidious/helpers/crystal_class_overrides.cr:34:5 in 'io'
  from /usr/share/crystal/src/http/client.cr:664:5 in 'send_request'
  from /usr/share/crystal/src/http/client.cr:599:5 in 'exec_internal_single'
  from /usr/share/crystal/src/http/client.cr:586:5 in 'exec_internal'
  from /usr/share/crystal/src/http/client.cr:581:7 in 'exec'
  from /usr/share/crystal/src/http/client.cr:706:5 in 'exec'
  from /usr/share/crystal/src/http/client.cr:738:7 in 'exec'
  from /usr/share/crystal/src/http/client.cr:406:3 in 'head'
  from src/invidious.cr:1514:7 in '->'
  from /usr/share/crystal/src/primitives.cr:266:3 in '->'
  from src/invidious/helpers/handlers.cr:30:5 in 'process_request'
  from lib/kemal/src/kemal/route_handler.cr:17:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from lib/kemal/src/kemal/websocket_handler.cr:13:14 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from lib/kemal/src/kemal/filter_handler.cr:21:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/handlers.cr:212:5 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/handlers.cr:94:12 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/handlers.cr:145:12 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/handlers.cr:68:12 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/static_file_handler.cr:189:11 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from lib/kemal/src/kemal/exception_handler.cr:8:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/logger.cr:17:35 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from lib/kemal/src/kemal/init_handler.cr:12:7 in 'call'
  from /usr/share/crystal/src/http/server/request_processor.cr:51:11 in 'process'
  from /usr/share/crystal/src/http/server.cr:515:5 in 'handle_client'
  from /usr/share/crystal/src/http/server.cr:468:13 in '->'
  from /usr/share/crystal/src/primitives.cr:266:3 in 'run'
  from /usr/share/crystal/src/fiber.cr:92:34 in '->'
  from ???
Caused by: SSL_connect: Success (RuntimeError)
```

</details>

File: src/invidious/yt_backend/youtube_api.cr
Comment (syeopite) [inline src/invidious/yt_backend/youtube_api.cr:414]: TODO: Check for the configuration option as well

**Review by SamantazFox — CHANGES_REQUESTED** [review]

File: src/invidious/routes/images.cr
Comment (SamantazFox) [inline src/invidious/routes/images.cr:253]: I'm wondering something: why isn't the `if CONFIG.use_quic` check made at the `YT_POOL` pool level, rather than here?

**Review by SamantazFox — CHANGES_REQUESTED** [review]

**Comment by syeopite** [comment]
Just added a commit to disable quic by default due to #2577. Once merged, assuming the user hasn't modified the `use_quic` attribute, Invidious should work ootb.

**Comment by unixfox** [comment]
For what is QUIC still used in the invidious code?

**Comment by syeopite** [comment]
QUIC shouldn't be used at all now when either `use_quic` is turned off, or the compile time flag `disable_quic` is applied. 

To be absolutely certain of that, you can just go with the latter route and remove the dependency entirely from the libs folder.

**Pull request merged**
